lib/tree.rb

```diff
@@ -16,43 +20,45 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log(n))
+  # Space Complexity: O(log(n))
+  def add(key, value = nil)
+    @root = add_under_root(key, value, @root)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log(n))
+  # Space Complexity: O(log(n))
   def find(key)
-    raise NotImplementedError
+    return find_under_root(key, @root)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    result = []
+    in_order_under_root(@root, result)
+    return result
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    return pre_order_under_root(@root)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    return post_order_under_root(@root)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(log(n))
   def height
```

Comment:
👍 The space complexity is O(n) if the tree is unbalanced and O(log n) if it is balanced.